contracts/compressor/IntentCompressor.sol

```diff
@@ -0,0 +1,335 @@
+    uint48 routeAmount;
+    // Prove type (INSTANT = 0, BATCH = 1)
+    uint8 proveType;
+    // Claimant
+    address claimant;
+}
+
+contract IntentCompressor {
+    using SafeERC20 for IERC20;
+
+    /**
+     * @notice Thrown when the vault has insufficient token allowance for reward funding
+     * @param token The token address
+     * @param spender The spender address
+     * @param amount The amount of tokens required
+     */
+    error InsufficientTokenAllowance(
+        address token,
+        address spender,
+        uint256 amount
+    );
+
+    address public immutable PROVER;
+    IInbox public immutable INBOX;
+    IIntentSource public immutable INTENT_SOURCE;
+
+    constructor(address _intentSource, address _inbox, address _prover) {
+        INBOX = IInbox(_inbox);
+        INTENT_SOURCE = IIntentSource(_intentSource);
+        PROVER = _prover;
+    }
+
+    function fulfill(
+        bytes32 payload,
+        bytes32 rewardHash,
+        bytes32 routeSalt
+    ) external returns (bytes[] memory) {
+        EncodedFulfillment memory encodedFulfillment = decodeFulfillPayload(
+            payload
+        );
+
+        Route memory route = _constructRoute(
+            routeSalt,
+            encodedFulfillment.sourceChainIndex,
+            encodedFulfillment.destinationChainIndex,
+            encodedFulfillment.routeTokenIndex,
+            encodedFulfillment.routeAmount
+        );
+
+        bytes32 routeHash = keccak256(abi.encode(route));
+        bytes32 intentHash = keccak256(abi.encodePacked(routeHash, rewardHash));
+
+        require(
+            route.tokens.length == 1,
+            "Cannot fulfill intent multiple tokens"
+        );
+
+        // Approve route token
+        TokenAmount memory routeToken = route.tokens[0];
+        IERC20(routeToken.token).approve(address(INBOX), routeToken.amount);
+
+        if (encodedFulfillment.proveType == 1) {
+            return
+                INBOX.fulfillHyperBatched(
+                    route,
+                    rewardHash,
+                    encodedFulfillment.claimant,
+                    intentHash,
+                    PROVER
+                );
+        } else {
+            return
+                INBOX.fulfillHyperInstant(
+                    route,
+                    rewardHash,
+                    encodedFulfillment.claimant,
+                    intentHash,
+                    PROVER
+                );
+        }
+    }
+
+    function publishTransferIntentAndFund(
+        bytes32 payload
+    ) external returns (bytes32 intentHash) {
+        EncodedIntent memory encodedIntent = decodePublishPayload(payload);
+        Intent memory intent = _constructIntent(encodedIntent);
+
+        _fundIntent(intent);
+
+        return INTENT_SOURCE.publishAndFund(intent, false);
+    }
+
+    // ======================== Public Functions ========================
+
+    function getChainIds() public pure returns (uint16[6] memory) {
```

Comment:
ChainID should probably be `uint256` instead of `uint16`. We use it in intent struct anyways and will save gas on conversions

Reply:
And here maybe also should be `if ... else if ... else` instead of an array.